### PR TITLE
6LoWPAN: Introduce test vectors and improvements

### PR DESCRIPTION
**Pull request by LuDuda**
This PR introduces a new 6LoWPAN test vectors (~60) and fixes some bugs.

Modification in the LoWPAN:
 - Fixed bug of performing stateful compression with context that has compression flag disable.
 - Fixed bug of decompression unspecified address "::"
 - Fixed bug of multicast address compression (48-bit)
 - Improve compression of Traffic Class and Flow Label fields (when DSCP is 0, and ECN is not)
 - Throw parse error when DAC=1 and DAM=00.
 - Add possibility to compress Unicast-Prefix-based IPv6 Multicast Addresses (however i've added support only if the address matches context 0 - for performance reason - but we can change it in the future)

An example of the unit test log:
```
=== Test name: Extension Header - Hop-by-Hop with PadN2 ===

Expected error -------------- no
UDP present ----------------- no
Extension Headers present --- yes
IP-in-IP present ------------ no
IPHC length ----------------- 13
IPv6 uncompressed offset ---- 48

Expected IPv6 uncompressed packet: 
60 00 00 00 00 10 00 40  fd 00 ca fe fa ce 12 34 
00 00 00 ff fe 00 00 00  ff 03 00 00 00 00 00 00 
00 00 00 00 00 00 00 01  3a 00 6d 02 60 11 01 00 
80 00 01 02 03 04 05 06  

Expected IPHC compressed frame: 
7e 7a 03 00 00 01 e0 3a  04 6d 02 60 11 80 00 01 
02 03 04 05 06 

Resulted IPHC compressed frame: 
7e 7a 03 00 00 01 e0 3a  04 6d 02 60 11 80 00 01 
02 03 04 05 06 

Resulted IPv6 uncompressed packet: 
60 00 00 00 00 10 00 40  fd 00 ca fe fa ce 12 34
00 00 00 ff fe 00 00 00  ff 03 00 00 00 00 00 00 
00 00 00 00 00 00 00 01  3a 00 6d 02 60 11 01 00 
80 00 01 02 03 04 05 06  

PASS
```

### CODE REVIEW

File: tests/unit/test_lowpan.cpp [hunk 1]
```diff
@@ -27,100 +27,1890 @@
...
+    Test(testVector, false, true);
+}
+
+static void TestErrorReservedValueDestination0100(void)
+{
+    TestIphcVector testVector("Reserved value of M-DAC-DAM - 0100");
+
+    // Setup MAC addresses.
+    testVector.SetMacSource(sTestMacSourceDefaultLong);
+    testVector.SetMacDestination(sTestMacDe
```
Comment (jwhui) [inline tests/unit/test_lowpan.cpp]: nit: prefer `Iphc` to `IPHC` and `Nhc` to `NHC`.

File: src/core/thread/lowpan.cpp [hunk 2]
```diff
@@ -194,7 +195,7 @@ int Lowpan::CompressMulticast(const Ip6::Address &aIpAddr, uint16_t &aHcCtl, uin
                 memcpy(cur + 1, aIpAddr.mFields.m8 + 13, 3);
                 cur += 4;
             }
-            else if (i >= 9)
+            else if (i >= 11)
```
Comment (turon) [inline src/core/thread/lowpan.cpp]: minor: Is there a way to make these IPv6 address offsets less magic?  
Even a short, targeted comment to align the compression states to the RFC text would be valued.

File: tests/unit/test_vector.c [hunk 3]
```diff
@@ -1,157 +0,0 @@
...
- */
-
-#include "test_vector.h"
-
-test_lowpan_vector_t sTestVectorLowpan[] =
-{
-    {
-        // I1_t1_AF_pass.pcap
-        .mTest = "LL64 unicast ICMP ping request",
-        .mCompressed =
```
Comment (turon) [inline tests/unit/test_vector.c]: Did any of these older vectors get carried forward to the new scheme, or have they all been supplanted with equivalent or better ones?

File: src/core/thread/lowpan.cpp (hunk 2 above)
Comment (LuDuda) [inline src/core/thread/lowpan.cpp]: Agree, i will add some comments to it.

File: tests/unit/test_vector.c (hunk 3 above)
Comment (jwhui) [inline tests/unit/test_vector.c]: @LuDuda, I like how the new tests all have a targeted purpose against the compressor/decompresssor, making coverage much easier to understand.  I would not worry about the legacy test vectors.

**Comment by LuDuda** [comment]
I've added comments as @turon requested and changed IPHC to Iphc and NHC to Nhc in function names.
Additional one more test have been added to fix one corner situation in TC and FL.
